File: 1_arrays_and_strings/1/main.c
// Grayson Pike, 2016
// Implement an algorithm to determine if a string has all unique characters.
// What if you cannot use additional data structures?

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

// Using no additional data structures
// very slow for large strings, exponential runtime
bool is_unique(char *s) {
    int len = strlen(s);
    for(int i = 0; i < len; i++) {
        for(int j = 0; j < len; j++) {
            if(s[i] == s[j] && i != j) {
                return false;
            }
        }
    }
    return true;
}

// using an array, assuming that ascii goes from 0-127
// linear runtime
bool is_unique2(char *s) {
    int len = strlen(s);
    bool used[128];
    memset(used, 0, sizeof(used));

    for(int i = 0; i < len; i++) {
        if(used[(int) s[i]]) {
            return false;
        } else {
            used[(int) s[i]] = true;
        }
    }
    return true;
}

int main() {

    char *s1 = "grayson";
    char *s2 = "riddhi";

    printf("Using is_unique:\n");
    printf("s1 unique: %s\n", is_unique(s1) ? "true" : "false");
    printf("s2 unique: %s\n", is_unique(s2) ? "true" : "false");

    printf("Using is_unique2:\n");
    printf("s1 unique: %s\n", is_unique2(s1) ? "true" : "false");
    printf("s2 unique: %s\n", is_unique2(s2) ? "true" : "false");

    return 0;
}

File: 1_arrays_and_strings/1/main.py
# Grayson Pike, 2016
# Implement an algorithm to determine if a string has all unique characters.
# What if you cannot use additional data structures?


# Using no additional data structures
# very slow for large strings, exponential runtime
def is_unique(s):
    for i in range(len(s)):
        for j in range(len(s)):
            if s[i] == s[j] and i != j:
                return False
    return True


# using an array, assuming that ascii goes from 0-127
# linear runtime
def is_unique2(s):
    used = []
    for i in range(128):
        used.append(False)
    for c in s:
        if used[ord(c)]:
            return False
        else:
            used[ord(c)] = True
    return True


def main():
    s1 = "grayson"
    s2 = "riddhi"

    print("Using is_unique:")
    print("s1 is unique: " + str(is_unique(s1)))
    print("s2 is unique: " + str(is_unique(s2)))

    print("Using is_unique2:")
    print("s1 is unique: " + str(is_unique2(s1)))
    print("s2 is unique: " + str(is_unique2(s2)))

main()
